src/compiler/checker.ts

```diff
@@ -10406,14 +10406,34 @@ namespace ts {
             const checkTypeInstantiable = maybeTypeOfKind(checkType, TypeFlags.Instantiable | TypeFlags.GenericMappedType);
             let combinedMapper: TypeMapper | undefined;
             if (root.inferTypeParameters) {
-                const context = createInferenceContext(root.inferTypeParameters, /*signature*/ undefined, InferenceFlags.None);
+                const freshParams = map(root.inferTypeParameters, cloneTypeParameter);
+                const freshMapper = createTypeMapper(root.inferTypeParameters, freshParams);
+                const context = createInferenceContext(freshParams, /*signature*/ undefined, InferenceFlags.None);
+                // We have three mappers that need applying:
+                //    * The original `mapper` used to create this conditional
+                //    * The mapper that maps the old root type parameter to the clone (`freshMapper`)
+                //    * The mapper that maps the clone to its inference result (`context.mapper`)
+                // When we're looking at making an inference for a clone, when we get its constraint, it's autmagically be
+                // instantiated with the context, so it doesn't need that one - however the constraint may refer to another _root_, _uncloned_
+                // `infer` type parameter.
+                // Eg, if we have `Foo<T, U extends T>` and `Foo<number, infer B>` - `B` is constrained to `T`, which, in turn, has been instantiated
+                // as `number`
+                // Conversely, if we have `Foo<infer A, infer B>`, `B` is still constrained to `T` and `T` is instantiated as `A`

```

Comment:
this sounds like two steps of unification. Is it guaranteed to be limited to two steps? What happens with the current code if it's not?

Reply:
We don't actually do unification so yeah, it is limited to two steps, since those are the two steps we actually do. If there were more, that'd either be incorporated into one of these mappers (so effectively a sub-step), or be represented by another mapper which'd need to be passed in and incorporated here.